[ENH] Add the ability to check the validity of an MAG

aryan26roy opened the pull request:
<!-- To ensure we can review your pull request promptly please complete this template entirely. -->

<!-- Please reference the issue number here. You can replace "Fixes" with "Closes" if it makes more sense. -->
Closes #90 

Changes proposed in this pull request:
<!-- Please list all changes/additions here. -->
- Add a function and related helper functions to check the validity of an MAG.

## Before submitting

<!-- Please complete this checklist BEFORE submitting your PR to speed along the review process. -->
- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

<!-- Please complete this checklist AFTER submitting your PR to speed along the review process. -->
- [ ] All GitHub Actions jobs for my pull request have passed.

aryan26roy commented:
@adam2392 any idea why the DCO is complaining? I have signed off my commits.

adam2392 commented:
> @adam2392 any idea why the DCO is complaining? I have signed off my commits.

I wouldn't worry about it. You can try doing the rebasing instructions to fix it if you want: https://github.com/py-why/pywhy-graphs/pull/91/checks?check_run_id=15461061662

aryan26roy commented:
@adam2392 In finding directed cycles, does the direction of the directed edge matter? 
For eg., is this a directed cycle?  a -> b -> c -> d <- a

adam2392 commented:
> @adam2392 In finding directed cycles, does the direction of the directed edge matter?
> For eg., is this a directed cycle? a -> b -> c -> d <- a

Sorry for the delay.

In causality, typically cycles are directed cycles. In directed cycles, by definition the direction matters. `a -> b -> c -> d <- a` is not a cycle, but `a -> b -> c -> d -> a` is.

I would review the [Zhang paper](https://www.sciencedirect.com/science/article/pii/S0004370208001008) if any questions like this arise in case I can't respond in a timely manner.

aryan26roy commented:
@adam2392 thanks for the clarification.

aryan26roy commented:
@adam2392 What should I have in the sets L and S when checking for an inducing path?

adam2392 commented:
> @adam2392 What should I have in the sets L and S when checking for an inducing path?

You can have them be passed in by user and default to empty sets.

aryan26roy commented:
@adam2392 I think the function is complete. Now I should add unit tests. Do you know how I can get examples of valid and invalid MAGs?

adam2392 commented:
> @adam2392 I think the function is complete. Now I should add unit tests. Do you know how I can get examples of valid and invalid MAGs?

I would just construct them by hand following the definition, or look at the papers on causal discovery (if you see them in a paper, cite where you got it from and which figure).

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +605,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
     return (path_exists, path)
+
+
+def _find_adc(G):
```
This can be a public function and tested separately to help you ensure that this works as intended.

aryan26roy replied:
Sorry for the late response, you're right. I'll do just that.

adam2392 commented on pywhy_graphs/algorithms/tests/test_generic.py:
```diff
@@ -228,3 +228,102 @@ def test_is_collider():
+
+
+def test_valid_mag():
+    # K -> H -> Z -> X -> Y -> J <- K
```
Can you add reasonings in the in-line comments for why these aren't valid MAGs?

aryan26roy replied:
Alright.

adam2392 commented on pywhy_graphs/algorithms/tests/test_generic.py:
```diff
@@ -228,3 +228,102 @@ def test_is_collider():
+
+
+def test_find_adc():
+    # K -> H -> Z -> X -> Y -> J <- K
```
Can you add reasonings in the in-line comments for why these are/aren't valid ADC?

aryan26roy replied:
Sure!

aryan26roy commented:
@adam2392 Is this good enough?

adam2392 commented:
Can you add a CHANGELOG entry to fix the CI?

And add these functions to `api.rst`?

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+        for elem in biedges:
+            if (elem[0] in ancestors and elem[1] in descendants) or (
+                elem[1] in ancestors and elem[0] in descendants
+            ):  # there is an undirected edge from one of the ancestors to a descendant
```
undirected or bidirected?

aryan26roy replied:
Ah error in the comments, should be bidirected.

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+
+def find_adc(G):
+
+    """Finds an Almost Directed Cycles in a mixed edge graph.
```
If you name it `find_*`, it should return the actual path found: E.g. see similar function in networkx: https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.find_cycle.html

Otw, you can name it `has_almost_directed_cycle`, which makes more sense. Or you can return the actual path that is the "almost directed cycle".

aryan26roy replied:
Got it.

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+    """
+
+    if L is None:
+        L = set()
```
This line is not tested.

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+        L = set()
+
+    if S is None:
+        S = set()
```
This line is not tested.

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+            if out[0] is True:
+                return False
+
+    return is_valid
```
just `return True` since `is_valid` is not used?

aryan26roy replied:
alright

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+    return adc_present
+
+
+def valid_mag(G: ADMG, L: set = None, S: set = None):
```
An additional check perhaps to add is that there is only one edge between any two nodes.

I.e. in a MAG, this is not possible: `G = (X -> Y; X <--> Y)`

aryan26roy replied:
I will also add a check for undirected edges (Since ADMG can have those).

aryan26roy commented:
@adam2392 I updated the CHANGELOG for v0.2, is that fine?

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,93 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+
+def valid_mag(G: ADMG, L: set = None, S: set = None):
+    """Checks if the provided graph is a valid MAG.
+
```
The docstring should describe what a "valid MAG" is.

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,93 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+
+def has_adc(G):
+    """Finds if a graph has an Almost Directed Cycle.
+
```
The docstring should describe what an "almost directed cycle" is

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,93 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+
+
+def has_adc(G):
+    """Finds if a graph has an Almost Directed Cycle.
```
```suggestion
    """Check if a graph has an almost directed cycle.
```

adam2392 reviewed (commented):
Few nits and then it LGTM. This will now help you finish https://github.com/py-why/pywhy-graphs/issues/85

Which will finally enable you to finish the PR for checking validity of a PAG perhaps(?).

We can always revisit this code if there is a bug you find

aryan26roy commented:
@adam2392 I added the definitions, can you take a look?

adam2392 commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,100 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+    """Check if a graph has an almost directed cycle.
+
+    An almost directed cycle is a is a directed cycle containing
+    one bidirected edge.
```
```suggestion
    one bidirected edge. For example, ``A -> B -> C <-> A`` is an adc.
```

adam2392 reviewed (approved):
LGTM after final few comments

aryan26roy commented on pywhy_graphs/algorithms/generic.py:
```diff
@@ -604,3 +606,100 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+
+    A valid MAG is a mixed edge graph that only has directed and bi-directed
+    edges, no directed or almost directed cycles and no inducing paths between
+    any two non-adjacent pair of nodes.
```
Sure

aryan26roy commented:
@adam2392 I rebased the branch and removed the quotes from my email address but the DCO won't stop complaining. Is there a way to merge this without DCO approval?

adam2392 commented:
Thanks @aryan26roy !
